### ISSUE TITLE
Order in which reifyWithLocals searches for Decs needs to be documented

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#131]
Pop quiz: what does this program do?

```hs
{-# LANGUAGE TemplateHaskell #-}
module Foo where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

type T = Char

$(pure [])

main :: IO ()
main = putStrLn
  $(do let t      = mkName "T"
           t_dec1 = TySynD t [] (ConT ''Int)
           t_dec2 = TySynD t [] (ConT ''Bool)
       t_info <- withLocalDeclarations [t_dec1] $
                 withLocalDeclarations [t_dec2] $
                 reifyWithLocals t
       stringE $ pprint t_info)
```

(a) It prints `type T = Char`.
(b) It prints `type T = Int`.
(c) It prints `type T = Bool`.
(d) It throws an error since there are multiple declarations in scope with the name `T`.

As it turns out, the answer is (a):

```
λ> main
type Foo.T = GHC.Types.Char
```

Now what happens if you comment out `type T = Char`? This time, it will print out:

```
λ> main
type T = GHC.Types.Int
```

This reveals an interesting behavior of the way `reifyWithLocals` behaves. It first looks through globally declared declarations (such as `type T = Char`). If it cannot find any matches, it then looks through local declarations, starting with the ones that were most recently added with `withLocalDeclarations`. Moreover, having a global and local declaration available with the same name (or even two local declarations with the same name) is _not_ an error, and adding such a declaration with `withLocalDeclarations` does not shadow existing ones.

I found this behavior somewhat surprising, and what's more, I couldn't find this behavior documented anywhere in the [Reification section](https://hackage.haskell.org/package/th-desugar-1.10/docs/Language-Haskell-TH-Desugar.html#g:7) of `L.H.T.Desugar`'s Haddocks. I think this deserves to be highlighted more prominently in the documentation.

**Comment by goldfirere** [comment]
Agreed that it should at least be documented. Though either erroring or shadowing would be better behavior than what you've demonstrated.

We should be careful, though: you've said "looks through local declarations, starting with the ones that were most recently added", but that would suggest that we get `Bool` when there is no top-level declaration for `T`.

**Comment by RyanGlScott** [comment]
One thing I omitted in https://github.com/goldfirere/th-desugar/issues/130#issue-559752978 is that we'll need to update the definition of `withLocalDeclarations`. Here is a sketch of what that might look like:

```hs
withLocalDeclarations :: DsMonad q => [Dec] -> DsM q a -> q a
withLocalDeclarations new_decs (DsM x) = do
  env <- ask
  let new_infos    = ...
      new_fixities = ...
      new_types    = ...
      env' = DsEnv { dsEnvDecs      = orig_decs ++ new_decs
                   , dsEnvDecMap    = combine (dsEnvDecMap    env) new_infos
                   , dsEnvFixityMap = combine (dsEnvFixityMap env) new_fixities
                   , dsEnvTypes     = combine (dsEnvTypeMap   env) new_types
                   }
  runReaderT x env'

combine :: Map Name a -> Map Name a -> Map Name a
combine old_map new_map = ???
```

Question: how should `combine` be implemented? Should it be this?

```hs
combine old_map new_map = Map.union old_map new_map
```

Or should it be this?

```hs
combine old_map new_map = Map.union new_map old_map
```

As it turns out, only the former version of `combine` will preserve the current semantics of `withLocalDeclarations`. Recall that `Map.union` is left biased, i.e., it prefers entries from the leftmost `Map` when duplicate keys are encountered. Therefore, `Map.union old_map new_map` matches what happens today in `withLocalDeclarations`, as adding a duplicate local declaration will not shadow the existing local declaration. (I only discovered this fact through experimentation, so this [ought to be documented more prominently](https://github.com/goldfirere/th-desugar/issues/131).)

**Comment by RyanGlScott** [comment]
> Agreed that it should at least be documented. Though either erroring or shadowing would be better behavior than what you've demonstrated.

Indeed. I wonder if anything would break (either in `singletons` or some other downstream `th-desugar` client) if we made adding duplicates an error...

> We should be careful, though: you've said "looks through local declarations, starting with the ones that were most recently added", but that would suggest that we get `Bool` when there is no top-level declaration for `T`.

Oops! I meant to write "_least_ recently added", I promise :)

**Comment by RyanGlScott** [comment]
To follow up on my own "I wonder if anything would break [...] if we made adding duplicates an error" question in https://github.com/goldfirere/th-desugar/issues/131#issuecomment-581950422, I performed an experiment when attempting to implement a fix for #130. I decided to see what would happen if (1) I made it an error to add local declarations (using `withLocalDeclarations`) that duplicate existing global or local declarations, and (2) compiled some non-trivial test cases that use `th-desugar` (e.g., the `singletons` library).

The short version is that nothing broke. The long answer is that all of these test cases only ever invoke `withLocalDeclarations` with quoted `Dec`s, which always bind fresh `Name`s to each new `Dec`. This effectively means that there is no risk of adding duplicates with this common approach. Technically, someone could hand-write some `Dec`s with already existing `Name`s and add them, but thankfully this almost never happens.

This raises the question: should we go ahead and making adding duplicates an error now? My original motivation for pursuing this question was in response to implementing a fix for #130, since adding duplicate declarations would actually have an effect on how caching reifications would work. However, I've since abandoned the design proposed in #130, since it ended up [not yielding any performance improvements](https://github.com/goldfirere/th-desugar/issues/130#issuecomment-585840755). As a result, my motivation for making adding duplicates an error has gone away.

We could still make adding duplicates an error in the current implementation of `withLocalDeclarations`, which just stores all local declarations as a giant `[Dec]`. This might actually prove to be annoying to accomplish, however, because searching for duplicates would require extra traversals of the giant `[Dec]`.

In light of this, my inclination is to document that the behavior of adding duplicate declarations via `withLocalDeclarations` is undefined. This gives us flexibility to change the implementation in the future while still giving a strong indication that adding duplicates is a Very Weird Thing™ that should be avoided if possible.
